### ISSUE TITLE
vst3 plain gui: accessibility of sliders

### ISSUE THREAD

**Issue by DavidBailes** [audacity/audacity#3039]
**Describe the bug**
Screen readers do not automatically read the the actual values of sliders which is shown in the text to the right of each slider

**To Reproduce**
1. On Windows, start the Narrator screen reader (Ctrl+Windows key+Enter).
2. Open a VST3 effect set to use the plain gui.
3. Tab to a slider, or with a slider as the focus, press Insert/Caps lock + up arrow to read the current control.

**Actual behavior**
The screen reader reads the name of the control and a percentage (1-100), but not the actual value shown by the text to the right of the slider.

**Expected behavior**
The accessibility name of the slider is the name of the control followed by the actual value of the slider, so that the value of the slider is automatically read when the slider gains focus, or when the user presses the keystroke to read the current control. This is implemented for VST plugins.

**Additional information (please complete the following information):**
 - OS: Windows 10
 - Version Audacity 3.2.0 alpha

**Additional context**
Regression on VST plugins

**Comment by dozzzzer** [comment]
Verifying on Windows 11: the issue can still be reproduced using the STRs above.
Once the parameter is in focus, the value reads correctly. Further on, when pressing Insert + Up arrow, it reads the percentage value.

https://user-images.githubusercontent.com/70063970/182220164-42c6f6ac-f464-4023-935f-5584349bd6b1.mp4

**Comment by DavidBailes** [comment]
> Further on, when pressing Insert + Up arrow, it reads the percentage value.

Let me clarify something in my steps to reproduce. In step 3, I should have said "press Narrator key + up arrow to read the current control", rather than "press Insert/Caps lock + up arrow to read the current control".
In Narrators settings the Narrator key can be set to Caps lock, Insert, or Caps lock or Insert (at least it can on Windows 10).
So with a slider as the focus, pressing Narrator key + up arrow should read the name, value, and percentage, just as it does when the slider gains focus.
Do you still think that there is a problem?

**Comment by dozzzzer** [comment]
Yes, mine is set to "Caps Lock or Insert":
<img width="300" alt="image" src="https://user-images.githubusercontent.com/70063970/182229479-95ad5fbc-4cc1-4b2e-b80d-7effa16070dd.png">


> should read the name, value, and **percentage**

I'm quite confused TBH.
In the original description of the ticket you're saying the expected behavior is:

> The accessibility name of the slider is the name of the control followed by the actual value of the slider

So should it be the _percentage_ or _the actual value_ of the slider?

**Comment by DavidBailes** [comment]
> Yes, mine is set to "Caps Lock or Insert": <img alt="image" width="300" src="https://user-images.githubusercontent.com/70063970/182229479-95ad5fbc-4cc1-4b2e-b80d-7effa16070dd.png">
> 
> > should read the name, value, and **percentage**
> 
> I'm quite confused TBH. In the original description of the ticket you're saying the expected behavior is:
> 
> > The accessibility name of the slider is the name of the control followed by the actual value of the slider
> 
> So should it be the _percentage_ or _the actual value_ of the slider?

When the slider gains focus, or when the user presses Narrator key + up arrow to read the current object, I would expect Narrator to read the accessibility name, followed by the value of the slider (that is, the percentage).
In this case the accessibility name is being set to the name of the slider (the visible label) followed by the actual value, so I would expect Narrator to read the visible label, followed the the actual value, followed by the percentage.

This may be the problem: when you are pressing Insert + up arrow, are you using the Insert key on the numerical keypad and with Num Lock on rather than off?

**Comment by DavidBailes** [comment]
> This may be the problem: when you are pressing Insert + up arrow, are you using the Insert key on the numerical keypad and with Num Lock on rather than off?

If this wasn't the problem, can you give me an example of exactly what narrator says when you press Insert + up arrow?

**Comment by dozzzzer** [comment]
> so I would expect Narrator to read the visible label, followed by the actual value, followed by the percentage.

It reads this as soon as a slider gains focus (i.e. a User tabs to it). But later on, it only reads the percentage value.

As can be seen from the video that I've attached, what it reads is this:

1. when the slider gains focus: <label> <the actual value> slider at <percentage> (example: "Output Gain, 1.05 dB, slider at 62"
2. then, when pressing **Narrator key (Insert/Caps Lock) + Up**: <percentage> (example: "62... 61... 61..." etc.)

Is that something that you'd expect?

**Comment by DavidBailes** [comment]
> Is that something that you'd expect?

No! It looks like the Narrator key is being ignored, and Audacity is receiving just the up arrow key.
An alternative keystroke to read the current line is Narrator key + I. What does that do?

**Comment by dozzzzer** [comment]
> What does that do?

It doesn't do anything, it reads  "I" or simply reads the percentage value when pressing the arrow up key. So it looks like it ignores Narrator key. This is on Windows 11 virtual machine I'm running on Mac (that might be important).

On Windows 10, though, it works just fine: when I press Narrator + I/arrow up, it reads the slider's current value, just the way you described it.

**Comment by DavidBailes** [comment]
> This is on Windows 11 virtual machine I'm running on Mac (that might be important).

I suspect that this is what is causing the problem. You could check this by trying these Narrator keystrokes in applications other than Audacity.

**Comment by dozzzzer** [comment]
@DavidBailes you're right. Apparently, it simply ignores the Narrator key.
@LWinterberg can I ask you to verify this on your Win 11 machine please?

**Comment by LWinterberg** [comment]
works fine for me

**Comment by DavidBailes** [comment]
> @DavidBailes you're right. Apparently, it simply ignores the Narrator key.

If you do a google search for something like "virtual machine windows insert key", you'll find this is a common issue, and has a number of solutions.
